pkg/kv/kvserver/loqrecovery/server.go

```diff
@@ -355,62 +373,227 @@ func (s Server) StagePlan(
+				if err != nil {
+					return err
+				}
+				res := resp.ResponsesByNodeID[nID]
+				if len(res.Infos) > 0 {
+					info = res.Infos[0]
+					return nil
+				}
+				return errors.Newf("range r%d not found on node n%d", rID, nID)
+			})
+		if err != nil {
+			return serverpb.RangeInfo{}, err
+		}
+		return info, nil
+	}
+
+	// Note that rangeCheckErr is a partial error, so we may have subset of ranges
+	// and an error, both of them will go to response.
+	unavailable, rangeCheckErr := func() ([]loqrecoverypb.RangeRecoveryStatus, error) {
+		var unavailable []loqrecoverypb.RangeRecoveryStatus
+		if req.MaxReportedRanges == 0 {
+			return nil, nil
+		}
+		err := contextutil.RunWithTimeout(ctx, "retrieve ranges health", retrieveKeyspaceHealthTimeout,
+			func(ctx context.Context) error {
+				start := keys.Meta2Prefix
+				for {
+					kvs, err := db.Scan(ctx, start, keys.MetaMax, rangeMetadataScanChunkSize)
+					if err != nil {
+						return err
+					}
+					if len(kvs) == 0 {
+						break
+					}
+					var endKey roachpb.Key
+					for _, rangeDescKV := range kvs {
+						endKey = rangeDescKV.Key
+						var d roachpb.RangeDescriptor
+						if err := rangeDescKV.ValueProto(&d); err != nil {
+							continue
+						}
+						h := checkRangeHealth(ctx, d, isNodeLive, getRangeInfo)
+						if h != loqrecoverypb.RangeHealth_Healthy {
+							if len(unavailable) >= int(req.MaxReportedRanges) {
+								return errors.Newf("found more failed ranges than limit %d",
+									req.MaxReportedRanges)
+							}
+							unavailable = append(unavailable, loqrecoverypb.RangeRecoveryStatus{
+								RangeID:     d.RangeID,
+								StartKey:    d.StartKey.AsRawKey(),
+								FailureType: h,
+							})
+						}
+					}
+					start = endKey.Next()
+				}
+				return nil
+			})
+		// Note: we are returning partial results and an error in case we time out
+		// or hit unavailability or scan limit
+		return unavailable, err
+	}()
+	rangeHealth := serverpb.RecoveryVerifyResponse_UnavailableRanges{
+		Ranges: unavailable,
+	}
+	if rangeCheckErr != nil {
+		rangeHealth.Error = rangeCheckErr.Error()
+	}
+
 	return &serverpb.RecoveryVerifyResponse{
-		Statuses: nss,
+		Statuses:                   nss,
+		DecommissionedNodeStatuses: decomStatus,
+		UnavailableRanges:          rangeHealth,
 	}, nil
 }
 
+func checkRangeHealth(
+	ctx context.Context,
+	d roachpb.RangeDescriptor,
+	liveFunc func(rd roachpb.ReplicaDescriptor) bool,
+	rangeInfo func(ctx context.Context, id roachpb.RangeID, nID roachpb.NodeID) (serverpb.RangeInfo, error),
+) loqrecoverypb.RangeHealth {
+	if d.Replicas().CanMakeProgress(liveFunc) {
+		return loqrecoverypb.RangeHealth_Healthy
+	}
+	log.Infof(ctx, "probing range %s as it can't make progress", d)
+	stuckReplica := false
+	healthyReplica := false
+	for _, r := range d.Replicas().Descriptors() {
+		// Check if node is in deleted nodes first.
+		log.Infof(ctx, "checking replica %s", r)
```

Comment:
Do we really need to log these and the other messages throughout? If we do, use vevents.